cdap-unit-test/src/test/java/co/cask/cdap/test/base/TestFrameworkTestBase.java

```diff
@@ -31,12 +43,65 @@
   public void afterTest() throws Exception {
     try {
       super.afterTest();
-      // Sleep a second before clear. There is a race between removal of RuntimeInfo
-      // in the AbstractProgramRuntimeService class and the clear() method, which loops all RuntimeInfo.
-      // The reason for the race is because removal is done through callback.
-      TimeUnit.SECONDS.sleep(1);
     } finally {
-      clear();
+      reset();
+    }
+  }
+
+  /**
+   * Creates an artifact jar by tracing dependency from the given {@link Application} class.
+   */
+  protected static File createArtifactJar(Class<? extends Application> appClass) throws IOException {
+    return new File(AppJarHelper.createDeploymentJar(new LocalLocationFactory(TMP_FOLDER.newFolder()),
+                                                     appClass).toURI());
+  }
+
+  /**
+   * Deploys an {@link Application} using the given artifact jar.
+   */
+  protected static ApplicationManager deployWithArtifact(Class<? extends Application> appClass,
+                                                         File artifactJar) throws Exception {
+    return deployWithArtifact(appClass, artifactJar, null);
+  }
+
+  /**
+   * Deploys an {@link Application} using the given artifact jar with an optional config object.
+   */
+  protected static <T> ApplicationManager deployWithArtifact(Class<? extends Application> appClass,
+                                                             File artifactJar, @Nullable T config) throws Exception {
+    ArtifactId artifactId = new ArtifactId(NamespaceId.DEFAULT.getNamespace(),
+                                           appClass.getSimpleName(), "1.0-SNAPSHOT");
+    addArtifact(artifactId, artifactJar);
+    AppRequest<T> appRequest = new AppRequest<>(new ArtifactSummary(artifactId.getArtifact(), artifactId.getVersion()),
+                                                config);
+    return deployApplication(NamespaceId.DEFAULT.app(appClass.getSimpleName()).toId(), appRequest);
+  }
+
+  protected void reset() {
+    // Retry clear() multiple times. There is a race between removal of RuntimeInfo
+    // in the AbstractProgramRuntimeService class and the clear() method, which loops all RuntimeInfo.
+    // The reason for the race is because removal is done through callback.
+    try {
+      int failureCount = 0;
+      Exception exception = null;
+      while (failureCount < 10) {
+        try {
+          exception = null;
+          clear();
+          break;
+        } catch (Exception e) {
+          exception = e;
+          failureCount++;
+          TimeUnit.MILLISECONDS.sleep(200);
+        }
+      }
+      
+      if (exception != null) {
+        throw exception;
+      }
+    } catch (Exception e) {
+      // If really fail to do reset, propagate the exception
+      throw Throwables.propagate(e);
```

Comment:
If the reset throws Exception in the signature, IntelliJ is not happy about that as the `reset` is called in a finally block.